File: codes/2023/2-First-BIT (Java)/src/main/java/cn/edu/bit/newnewcc/pass/ir/structure/LoopClone.java
package cn.edu.bit.newnewcc.pass.ir.structure;

import cn.edu.bit.newnewcc.ir.Value;
import cn.edu.bit.newnewcc.ir.exception.CompilationProcessCheckFailedException;
import cn.edu.bit.newnewcc.ir.value.BasicBlock;
import cn.edu.bit.newnewcc.ir.value.Instruction;
import cn.edu.bit.newnewcc.ir.value.instruction.JumpInst;
import cn.edu.bit.newnewcc.ir.value.instruction.PhiInst;

import java.util.*;

public class LoopClone extends CloneBase {

    private final BasicBlock headBlock;
    private final Collection<BasicBlock> loopBasicBlocks;

    private final List<BasicBlock> clonedBasicBlocks = new ArrayList<>();
    private final Map<PhiInst, PhiInst> headPhiMap = new HashMap<>();
    private final BasicBlockSymbol nextLoopHeadSymbol = new BasicBlockSymbol();

    public LoopClone(BasicBlock headBlock, Collection<BasicBlock> loopBasicBlocks) {
        this.headBlock = headBlock;
        this.loopBasicBlocks = loopBasicBlocks;
        doLoopClone();
    }

    public void setNextLoopHead(BasicBlock nextLoopHead) {
        nextLoopHeadSymbol.replaceAllUsageTo(nextLoopHead);
    }

    public void addEntry(BasicBlock basicBlock, Map<PhiInst, Value> valueMap) {
        if (!valueMap.keySet().equals(headPhiMap.keySet())) {
            throw new CompilationProcessCheckFailedException();
        }
        valueMap.forEach((entryPhi, entryValue) -> {
            headPhiMap.get(entryPhi).addEntry(basicBlock, entryValue);
        });
    }

    public BasicBlock getClonedLoopHead() {
        return (BasicBlock) getReplacedValue(headBlock);
    }

    public List<BasicBlock> getClonedBasicBlocks() {
        return Collections.unmodifiableList(clonedBasicBlocks);
    }

    public Value getClonedValue(Value originalValue) {
        return getReplacedValue(originalValue);
    }

    private void doLoopClone() {
        // 构建占位符
        for (BasicBlock basicBlock : loopBasicBlocks) {
            addBasicBlock(basicBlock);
            for (Instruction instruction : basicBlock.getInstructions()) {
                addSymbolFor(instruction);
            }
        }
        // 构建函数体
        for (BasicBlock basicBlock : loopBasicBlocks) {
            var clonedBlock = (BasicBlock) getReplacedValue(basicBlock);
            for (Instruction instruction : basicBlock.getInstructions()) {
                Instruction clonedInstruction;
                if (basicBlock == headBlock && instruction instanceof PhiInst) {
                    clonedInstruction = new PhiInst(instruction.getType());
                    headPhiMap.put((PhiInst) instruction, (PhiInst) clonedInstruction);
                } else if (instruction instanceof JumpInst jumpInst && jumpInst.getExit() == headBlock) {
                    clonedInstruction = new JumpInst(nextLoopHeadSymbol);
                } else {
                    clonedInstruction = cloneInstruction(instruction);
                }
                setValueMapKv(instruction, clonedInstruction);
                clonedBlock.addInstruction(clonedInstruction);
            }
            this.clonedBasicBlocks.add(clonedBlock);
        }
    }

}

File: codes/2024/25-Excellence-BNU-99-21-43 (C++)/workspace/source/ir/pass/loopclone.cpp
#include "loopclone.h"

namespace IR
{

    LoopClone::LoopClone(BasicBlock *headBlock, const std::set<BasicBlock *> &loopBasicBlocks)
        : headBlock(headBlock), loopBasicBlocks(loopBasicBlocks)
    {
        nextLoopHeadSymbol = BasicBlock::get("nextLoopHeadSymbol");
        doLoopClone();
    }

    void LoopClone::setNextLoopHead(BasicBlock *nextLoopHead)
    {
        nextLoopHeadSymbol->replaceAllUsageTo(nextLoopHead);
    }

    void LoopClone::addEntry(BasicBlock *basicBlock, const std::map<PhiInstruction *, Value *> &valueMap)
    {
        if (valueMap.size() != headPhiMap.size())
        {
            Error::Error(__PRETTY_FUNCTION__, __LINE__, "Invalid value map size");
        }
        for (const auto &[entryPhi, entryValue] : valueMap)
        {
            auto it = headPhiMap.find(entryPhi);
            if (it != headPhiMap.end())
            {
                it->second->insertIncomingValue(std::make_pair(entryValue, basicBlock));
            }
        }
    }

    BasicBlock *LoopClone::getClonedLoopHead() const
    {
        return static_cast<BasicBlock *>(getReplacedValue(headBlock));
    }

    const std::vector<BasicBlock *> &LoopClone::getClonedBasicBlocks() const
    {
        return clonedBasicBlocks;
    }

    Value *LoopClone::getClonedValue(Value *originalValue) const
    {
        return getReplacedValue(originalValue);
    }

    void LoopClone::doLoopClone()
    {
        // 构建占位符
        for (BasicBlock *basicBlock : loopBasicBlocks)
        {
            addBasicBlock(basicBlock);
            for (Instruction *instruction : basicBlock->getVectorInstructions())
            {
                addSymbolFor(instruction);
            }
        }

        // 构建函数体
        for (BasicBlock *basicBlock : loopBasicBlocks)
        {
            BasicBlock *clonedBlock = static_cast<BasicBlock *>(getReplacedValue(basicBlock));
            for (Instruction *instruction : basicBlock->getVectorInstructions())
            {
                // instruction->emitIR(std::cerr);
                Instruction *clonedInstruction;
                if (basicBlock == headBlock && dynamic_cast<PhiInstruction *>(instruction))
                {
                    clonedInstruction = PhiInstruction::create(instruction->getType(), std::vector<PVB>());
                    headPhiMap[static_cast<PhiInstruction *>(instruction)] = static_cast<PhiInstruction *>(clonedInstruction);
                }
                else if (auto jumpInst = dynamic_cast<BranchInstruction *>(instruction); jumpInst && jumpInst->isUnconditional() && jumpInst->getUnconditionalBlock() == headBlock)
                {
                    clonedInstruction = BranchInstruction::createBr(nextLoopHeadSymbol);
                }
                else
                {
                    clonedInstruction = cloneInstruction(instruction);
                }
                setValueMapKv(instruction, clonedInstruction);
                clonedBlock->InsertInstructionBack(clonedInstruction);
            }
            clonedBasicBlocks.push_back(clonedBlock);
        }
    }

}